src/umap/store/SparseStore.cpp

```diff
@@ -114,32 +134,59 @@ namespace Umap {
           if (close_status != 0){
             UMAP_LOG(Warning,"SparseStore: Failed to close file with id: " << i << " - " << strerror(errno));
           }
-         return_status = return_status | close_status;
-	}
+          return_status = return_status | close_status;
+        }  
       }
       return return_status;
     }
 
+    size_t SparseStore::get_current_capacity(){
+      return current_capacity;
+    }
+    
+    /**
+     * To get the size of any persistent region created using SparseStore without the need to instianiate an object
+    **/ 
+    size_t SparseStore::get_capacity(std::string base_path){
+      size_t capacity = 0;
+      std::string metadata_path = base_path + "/_metadata";
+      std::ifstream metadata(metadata_path.c_str());
+      if (!metadata.is_open()){
+        UMAP_ERROR("Failed to open metadata file" << " - " << strerror(errno));
+      }
+      else{
+        metadata.ignore ( std::numeric_limits<std::streamsize>::max(), '\n' );
+        metadata >> capacity;
+      }
+      return capacity;
+    }
+
     int SparseStore::get_fd(off_t offset, off_t &file_offset){
       int fd_index = offset / file_size;
       file_offset = offset % file_size; 
       std::string filename = root_path + "/" + std::to_string(fd_index);
       if ( file_descriptors[fd_index].id == -1 ){
             creation_mutex.lock(); // Grab mutex (only in case of creating new file, rather than serializing a larger protion of the code)
             if (file_descriptors[fd_index].id == -1){ // Recheck the value to make sure that another thread did not already create the file
-              int fd;
-              if ((fd = open(filename.c_str(), O_CREAT | O_RDWR | O_LARGEFILE | O_DIRECT, S_IRUSR | S_IWUSR)) != -1){
-                int fallocate_status;
-                if( (fallocate_status = posix_fallocate(fd,0,file_size) ) == 0){
-                  file_descriptors[fd_index].id = fd;
-                }
-                else{
-                  UMAP_ERROR("SparseStore: fallocate() failed for file with id: " << fd_index << " - " << strerror(errno));
+              int flags = (read_only ? O_RDONLY :  O_RDWR ) | O_CREAT | O_DIRECT | O_LARGEFILE;
+              int fd = open(filename.c_str(), flags, S_IRUSR | S_IWUSR);
+              if (fd == -1){
+                // retry without O_DIRECT
+                flags = (read_only ? O_RDONLY :  O_RDWR ) | O_CREAT | O_LARGEFILE;
```

Comment:
Can you explain when this retry is needed?

Reply:
Some file systems do not support "O_DIRECT" (e.g. TMPFS). So it first tries open() with O_DIRECT, and if fails, it retries without O_DIRECT to eliminate this possibility. If still fails, it throws an error.

Reply:
Consider adding a here? "Supporting FS that do not support O_DIRECT, e.g., TMPFS"